<a href="https://colab.research.google.com/github/masonnystrom/DS-Unit-1-Sprint-2-Statistics/blob/master/DS-Unit-1-Sprint-2-Statistics/module1/Notes_LS_DS13_121_Statistics_Probability_and_Inference.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>

<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Lesson 1*

# Statistics, Probability and Inference


## What is Descriptive Statistics?

<https://statistics.laerd.com/statistical-guides/descriptive-inferential-statistics.php>

In [2]:
import pandas as pd
df = pd.DataFrame({'a': [1,2,3,4,5], 'b': [2,4,6,8,10]})
df.head()

a   b
0  1   2
1  2   4
2  3   6
3  4   8
4  5  10

In [3]:
# How can we quickly look at some descriptive statistics of the above dataframe?
df.describe()

a          b
count  5.000000   5.000000
mean   3.000000   6.000000
std    1.581139   3.162278
min    1.000000   2.000000
25%    2.000000   4.000000
50%    3.000000   6.000000
75%    4.000000   8.000000
max    5.000000  10.000000

## What is Inferential Statistics?

<https://slideplayer.com/slide/5130463/16/images/2/Statistical+Inference.jpg>

## Hypothesis Testing (T-Tests)



### One Sample T-test

Is the sample reflective of our concept of the population?

In [0]:
# TODO - during class, but please help!
import numpy as np

In [7]:
coinflips = np.random.binomial(n=1, p=.5, size=12)

coinflips.mean()

0.6666666666666666

the population is all the potential coinflips using a fair coin where prob = .5

To fight against randomness we collect more data 

In [13]:
# mean 
coinflips = np.random.binomial(n=1, p=.5, size=200)

coinflips.mean()

0.515

Increase in sample size generates a more prcise mean and allows us to make inferences with greater confidence 

In [27]:
# non fair coin
# ryan gives 2 dollars to tyler for heads and gets 1 dollar for tails
coinflips = np.random.binomial(n=1, p=.8, size=20)
coinflips

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 1])

In [23]:
# ryan gives tyler
coinflips.sum()*2

32

In [24]:
# Tyler gives ryan
len(coinflips) - coinflips.sum()

4

In [28]:
coinflips = np.random.binomial(n=1, p=.8, size=50)
sample_mean = coinflips.mean()
sample_mean

0.94

# T-test


In [0]:
from scipy import stats 


In [30]:
# pass this function to my sample and what I think the pop is
# this will run my t-test
stats.ttest_1samp(coinflips, .5)

Ttest_1sampResult(statistic=12.96913958442439, pvalue=1.8386698234210102e-17)

In [0]:
"""test statistic = 12.96
p value = 0.000000183
the p value is the probability of obtaining a sample similar
to the one that we collected if the null hypothesis is true 

If htis is a fair coin, there's a .000000183 chance that we would have collected 
this sample via a random sampling

Conclusion: based on a t-statistic of 12.96 and a p value of .000000183 we reject
the null hypothesis that this is a fair coin and suggest the alternative that it is an
unfair coin. 
"""

### Load and Clean the Data

In [0]:
import pandas as pd

In [5]:
# Grab the file from UCI
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data

--2020-02-17 17:22:47--  https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18171 (18K) [application/x-httpd-php]
Saving to: ‘house-votes-84.data’

house-votes-84.data 100%[===================>]  17.75K  --.-KB/s    in 0.06s   

2020-02-17 17:22:47 (276 KB/s) - ‘house-votes-84.data’ saved [18171/18171]



In [32]:
df = pd.read_csv('house-votes-84.data', names=['party','handicapped-infants','water-project',
                          'budget','physician-fee-freeze', 'el-salvador-aid',
                          'religious-groups','anti-satellite-ban',
                          'aid-to-contras','mx-missile','immigration',
                          'synfuels', 'education', 'right-to-sue','crime','duty-free',
                          'south-africa'])

df.head()

party handicapped-infants water-project  ... crime duty-free south-africa
0  republican                   n             y  ...     y         n            y
1  republican                   n             y  ...     y         n            ?
2    democrat                   ?             y  ...     y         n            n
3    democrat                   n             y  ...     n         n            y
4    democrat                   y             y  ...     y         y            y

[5 rows x 17 columns]

In [34]:
df = df.replace({"y": 1, 'n':0, "?": np.NaN})

df.head()

party  handicapped-infants  ...  duty-free  south-africa
0  republican                  0.0  ...        0.0           1.0
1  republican                  0.0  ...        0.0           NaN
2    democrat                  NaN  ...        0.0           0.0
3    democrat                  0.0  ...        0.0           1.0
4    democrat                  1.0  ...        1.0           1.0

[5 rows x 17 columns]

In [35]:
rep = df[df['party']=='republican']
rep.head()

party  handicapped-infants  ...  duty-free  south-africa
0   republican                  0.0  ...        0.0           1.0
1   republican                  0.0  ...        0.0           NaN
7   republican                  0.0  ...        NaN           1.0
8   republican                  0.0  ...        0.0           1.0
10  republican                  0.0  ...        0.0           0.0

[5 rows x 17 columns]

In [36]:
dem = df[df['party']=='democrat']
dem.head()

party  handicapped-infants  water-project  ...  crime  duty-free  south-africa
2  democrat                  NaN            1.0  ...    1.0        0.0           0.0
3  democrat                  0.0            1.0  ...    0.0        0.0           1.0
4  democrat                  1.0            1.0  ...    1.0        1.0           1.0
5  democrat                  0.0            1.0  ...    1.0        1.0           1.0
6  democrat                  0.0            1.0  ...    1.0        1.0           1.0

[5 rows x 17 columns]

In [37]:
df['party'].value_counts()

democrat      267
republican    168
Name: party, dtype: int64

### Before Running the Test

In [38]:
"""
republican support fo the handicapped-infants bill
Null Hypothesis: Republicans suport this bill 50/50 or that the
party is split in their support for this bill. mu==.5
Alternative Hypothesis: The party is not split in support for this bill

confidence level of 95% and critical value of .05 (1-.95)
"""

'\nrepublican support fo the handicapped-infants bill\nNull Hypothesis: Republicans suppor this bill 50/50 or that the\nparty is split in their support for this bill. mu==.5\nAlternative Hypothesis: The party is not split in support for this bill\n'

# run the test

In [41]:
# only considering reps who actually voted 
stats.ttest_1samp(rep['handicapped-infants'], .5, nan_policy='omit')

Ttest_1sampResult(statistic=-10.232833482397659, pvalue=2.572179359890009e-19)

### After Running the Test


based on a t stat of 10.23 and a p-value of 0.0000000257, I reject the null hypothsis that support for the bill was split. And suggest the alternative that support is something other than 50%  

p value represetns that I would have observed this data with as ample if the null hypothsis was true. 

In [42]:
rep['handicapped-infants'].mean()

0.18787878787878787

### Conclusion

### Two Sample T-test

Comparing two groups. Are they similar or different?

1) Null Hypothesis: Democrat support for the 'education' bill is equal to republican support for the 'education bill

𝐻0:𝑥¯𝑟==𝑥¯𝑑 

This is the only kind of null hypothesis with a two sample t-test. Are the sample means the same or are they different?

2) Alternative Hypothesis: Democrat support for the 'education' bill is not equal to republican support for the 'education bill

𝐻0:𝑥¯𝑟≠𝑥¯𝑑 

3) Confidence Level: 95%

In [43]:
stats.ttest_ind(rep['education'], dem['education'], nan_policy='omit')

Ttest_indResult(statistic=20.500685724563073, pvalue=1.8834203990450192e-64)

Conclusion: based on a t-statistic of -20 and p value of 0. I reject the null hypothesis that republicans and democrats support the 'education' bill at the same level/rate. And I suggest the alternative that democrat and republican levels of support for this bill are different.

In [45]:
print(rep['education'].mean())
print(dem['education'].mean())

0.8709677419354839
0.14457831325301204




Ever thought about how long it takes to make a pancake? Have you ever compared the tooking time of a pancake on each eye of your stove? Is the cooking time different between the different eyes? Now, we can run an experiment and collect a sample of 1,000 pancakes on one eye and another 800 pancakes on the other eye. Assumed we used the same pan, batter, and technique on both eyes. Our average cooking times were 180 (5 std) and 178.5 (4.25 std) seconds repsectively. Now, we can tell those numbers are not identicial, but how confident are we that those numbers are practically the same? How do we know the slight difference isn't caused by some external randomness?

Yes, today's lesson will help you figure out how long to cook your pancakes (*theoretically*). Experimentation is up to you; otherwise, you have to accept my data as true. How are going to accomplish this? With probability, statistics, inference and maple syrup (optional). 

<img src="https://images.unsplash.com/photo-1541288097308-7b8e3f58c4c6?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=3300&q=80" width=400>



In [0]:
mu1 = 180 # sample mean
sigma1 = 5 # standard deviation

sample1 = np.random.normal(mu1, sigma1, 1000)

In [0]:
mu2 = 178.5 # sample mean
sigma2 = 4.25 # standard deviation

sample2 = np.random.normal(mu2, sigma2, 800)

Null H: the two burners cook pancakes at the same rate
Alt H: the burners cook at diff rates
conf: 95%

In [48]:
stats.ttest_ind(sample1, sample2)

Ttest_indResult(statistic=5.802172089251842, pvalue=7.718977864640943e-09)

Based on a t stat of 5.8 and a p value of .00000007718 we reject teh null hypothesis that the burners cook pancakes at the same rate and suggest the alternative that they cook at a different rate. 

the difference in pancake cooking times between the two burners is not solely due to the  smapleing process but that there is something inherently diff in the burners. 

## Understanding the Test Statistic and P-value

https://homepage.divms.uiowa.edu/~mbognar/applets/t.html

![T Statistic Equation](http://lambdachops.com/img/t-statistic-equation.png)

# Resources

- https://homepage.divms.uiowa.edu/~mbognar/applets/t.html
- https://rpsychologist.com/d3/tdist/
- https://gallery.shinyapps.io/tdist/
- https://en.wikipedia.org/wiki/Standard_deviation#Sample_standard_deviation_of_metabolic_rate_of_northern_fulmars
- https://www.khanacademy.org/math/ap-statistics/two-sample-inference/two-sample-t-test-means/v/two-sample-t-test-for-difference-of-means

# Why do we use the t-distribution in hypothesis tests?

![t-distribution-low-dof](https://lambdachops.com/img/t-distribution-low-dof.png)

![t-distribution-approximates-normal](https://github.com/ryanallredblog/ryanallredblog.github.io/blob/master/img/t-distribution-approximates-normal.png?raw=true)

### Helpful video on why we use the t-distribution

<https://www.youtube.com/watch?v=Uv6nGIgZMVw>

However, in order to understand it you'll need to understand what a z-score is:

A z-score calculates the number of standard deviations an observations lies from the population mean. The problem is that in real-world situations, we don't know what the sample mean is, so we have to turn to using the sample mean to estimate the population mean. Because the sample mean is generated from a sample and used to estimate the population mean with some level of uncertainty, it also has its own distribution a nd spread. This means that for low sample sizes both our estimates of the sample mean and sample population are not very precise, they're kind of spread out. It's this spread that makes the t-distribution wider than the normal distribution for low sample sizes. However, with the larger the sample size, the closer the t-distribution approximates the normal distribution.
